cli_tools/google-osconfig-agent/ospatch/patch_run.go

```diff
@@ -284,75 +269,91 @@ func (r *patchRun) runPatch() {
 	if err := r.createClient(); err != nil {
 		logger.Errorf("Error creating osconfig client: %v", err)
 	}
-	defer r.close()
+	defer r.complete()
 
 	for {
 		switch r.PatchStep {
 		default:
-			r.finishAndReportError(fmt.Sprintf("Unknown step: %q", r.PatchStep))
+			r.reportFailedState(fmt.Sprintf("unknown step: %q", r.PatchStep))
 			return
 		case acked:
 			logger.Debugf("Starting patchJob %s", r.Job)
 			r.StartedAt = time.Now()
-			r.PatchStep = prePatchReboot
+			if err := r.setStep(prePatchReboot); err != nil {
+				r.reportFailedState(fmt.Sprintf("Error saving agent step: %v", err))
+			}
 
-			if r.reportState(osconfigpb.Instance_STARTED) {
+			if err := r.reportContinuingState(osconfigpb.Instance_STARTED); err != nil {
+				r.handleErrorState(err.Error(), err)
 				return
 			}
 		case prePatchReboot:
-			if r.prePatchReboot() {
+			// We do this now to avoid a reboot loop prior to patching.
+			if err := r.setStep(patching); err != nil {
+				r.reportFailedState(fmt.Sprintf("Error saving agent step: %v", err))
+			}
+			if err := r.prePatchReboot(); err != nil {
+				r.handleErrorState(fmt.Sprintf("Error runnning prePatchReboot: %v", err), err)
 				return
 			}
-			r.PatchStep = patching
 		case patching:
 			if r.Job.DryRun {
-				if r.reportState(osconfigpb.Instance_APPLYING_PATCHES) {
+				if err := r.reportContinuingState(osconfigpb.Instance_APPLYING_PATCHES); err != nil {
+					r.handleErrorState(err.Error(), err)
 					return
 				}
 				logger.Infof("Dry run - No updates applied for patch job '%s'", r.Job.PatchJob)
 			} else {
 				if err := runUpdates(r); err != nil {
-					r.finishAndReportError(fmt.Sprintf("Failed to apply patches: %v", err))
+					r.handleErrorState(fmt.Sprintf("Failed to apply patches: %v", err), err)
 					return
 				}
 			}
-			if r.postPatchReboot() {
+			if err := r.postPatchReboot(); err != nil {
+				r.handleErrorState(fmt.Sprintf("Error runnning postPatchReboot: %v", err), err)
 				return
 			}
-			// We have not rebooted so no need to rerun updates.
-			r.PatchStep = completed
-		case completed:
-			r.reportSucceeded()
-			logger.Debugf("Completed patchJob %s", r.Job)
+			// We have not rebooted so patching is complete.
+			if err := r.setStep(complete); err != nil {
+				r.reportFailedState(fmt.Sprintf("Error saving agent step: %v", err))
+			}
+		case complete:
+			isFinalRebootRequired, err := systemRebootRequired()
+			if err != nil {
+				r.reportFailedState(fmt.Sprintf("Unable to check if reboot is required: %v", err))
```

Comment:
nit: "Unable..." is somewhat odd and normally your error messages start with "Failed to ..." or "Error *ing...". Should this be consistent with those?